lib/Analysis/Allocation.cpp

```diff
@@ -595,45 +715,73 @@ class AllocationAnalysis {
                 DenseMap<BufferT *, size_t> &bufferStart) {
     // Reset shared memory size
     allocation->sharedMemorySize = 0;
+
+    SmallVector<BufferT *> xBuffers;
+    for (auto *buf : buffers) {
+      if (buf->kind != BufferT::BufferKind::Alias)
+        xBuffers.push_back(buf);
+    }
+
     // First-fit graph coloring
     // Neighbors are nodes that interfere with each other.
     // We color a node by finding the index of the first available
     // non-neighboring node or the first neighboring node without any color.
     // Nodes with the same color do not interfere with each other.
     DenseMap<BufferT *, int> colors;
     for (auto value : buffers) {
-      colors[value] = (value == buffers[0]) ? 0 : -1;
+      colors[value] = (value == buffers.front()) ? 0 : -1;
     }
     SmallVector<bool> available(buffers.size());
-    for (auto x : buffers) {
+    for (auto x : xBuffers) {
       std::fill(available.begin(), available.end(), true);
       for (auto y : interference.lookup(x)) {
-        int color = colors[y];
-        if (color >= 0) {
-          available[color] = false;
-        }
+        y->apply([&](BufferT *buf) {
+          if (buf != x) {
+            int color = colors[buf];
+            if (color >= 0)
+              available[color] = false;
+          }
+        });
       }
-      auto it = std::find(available.begin(), available.end(), true);
+      auto it = llvm::find(available, true);
       colors[x] = std::distance(available.begin(), it);
+      LLVM_DEBUG(llvm::dbgs()
+                 << "    COLOR: " << x->id << " -> " << colors[x] << "\n");
     }
     // Finalize allocation
     // color0: [0, 7), [0, 8), [0, 15) -> [0, 7), [0, 8), [0, 15)
     // color1: [7, 9) -> [0 + 1 * 15, 9 + 1 * 15) -> [15, 24)
     // color2: [8, 12) -> [8 + 2 * 15, 12 + 2 * 15) -> [38, 42)
     // TODO(Keren): We are wasting memory here.
     // Nodes with color2 can actually start with 24.
-    for (auto x : buffers) {
+    for (auto x : xBuffers) {
       size_t adj = 0;
       for (auto y : interference.lookup(x)) {
-        adj = std::max(adj, bufferStart.lookup(y) + y->size);
+        y->apply([&](BufferT *buf) {
+          if (buf != x)
+            adj = std::max(adj, buf->size);
```

Comment:
bump only by max buffer->size.. should this be x->alignment until it finds an open slot?